scikits/odes/sundials/cvode.pyx

```diff
@@ -1180,22 +1205,66 @@ cdef class CVODE:
         np_tspan = np.asarray(tspan, dtype=float)
         np_y0    = np.asarray(y0, dtype=float)
 
-        return self._solve(np_tspan, np_y0)
+        soln = self._solve(np_tspan, np_y0)
+
+        flag_no, t, y, t_err, y_err = soln[:5]
+
+        if soln[5]:
+            t_roots = np.array(soln[5])
+        else:
+            t_roots = None
+
+        if soln[6]:
+            y_roots = np.array(soln[6])
+        else:
+            y_roots = None
+
+        if soln[7]:
+            t_tstop = np.array(soln[7])
+        else:
+            t_tstop = None
+
+        if soln[8]:
+            y_tstop = np.array(soln[8])
+        else:
+            y_tstop = None
+
+        flag = self.RETURN_FLAGS.get(flag_no, "UNKNOWN")
+
+        soln = CVODE_Return(
+            flag=flag, flag_no=flag_no, values=(t,y), err_values=(t_err,y_err),
+            root_values=(t_roots,y_roots), tstop_values=(t_tstop,y_tstop)
+        )
+
```

Comment:
I tried avoiding putting too much logic in the _solve function for the return values, because if we add more stuff to solve, that logic may get in the way. One-liners would have been better.

With (1), I had a version which did that for some of the flags ages ago (even including values where it gave more context). I removed it as cvode already provided the same info in their output, which can controlled using err_handler. If you think it's valuable to have that, then I can add it in.

For some of the stuff I work on, I have some really horrible equations, which in theory if the initial conditions are slightly off do crazy things (I still haven't got them working...). I keep the flag_name (easier to remember than the int value) so I know how it failed and can compare with previous results via graphing etc. I like IntEnum idea, given that's effectively what cvode does.

Currently, if interruptfn is not used, then if either a root or tstop is encountered, then it's returned inside soln.error, as I didn't change that (due to oversight, not any kind of plan).